scripts/ag

```diff
@@ -105,32 +147,118 @@ def clean_test():
     cur = conn.cursor()
 
+    cur.execute('SET SEARCH_PATH TO ag, public')
+
     conn.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)
 
-    if is_test_db == 'false':
-        raise OperationalError("The settings table indicates this is not "
-                               "a test database; aborting.")
-    else:
-        cur.execute('drop database {}'.format(AMGUT_CONFIG.database))
-        print("Database %s dropped" % AMGUT_CONFIG.database)
-        cur.execute('drop role ag_wwwuser')
-        print("role ag_wwwuser dropped")
+    click.echo("Dropping user ag_wwwuser")
+
+    try:
+        click.echo('Reassigning owned objects')
+        cur.execute('REASSIGN OWNED BY ag_wwwuser TO {}'
+                    .format(AMGUT_CONFIG.user))
+
+        # Database
+        click.echo('Revoking all privileges on {}'
+                   .format(AMGUT_CONFIG.database))
+        cur.execute('REVOKE ALL PRIVILEGES ON DATABASE {} '
+                    'FROM ag_wwwuser CASCADE'.format(AMGUT_CONFIG.database))
+
+        # public schema
+        click.echo('Revoking all privileges in schema public')
+        cur.execute('REVOKE ALL PRIVILEGES ON SCHEMA public FROM ag_wwwuser '
+                    'CASCADE')
+
+        click.echo('Revoking all privileges on all tables in public')
+        cur.execute('REVOKE ALL PRIVILEGES ON ALL TABLES IN SCHEMA public '
+                    'FROM ag_wwwuser CASCADE')
+
+        click.echo('Revoking all privileges on all functions in public')
+        cur.execute('REVOKE ALL PRIVILEGES ON ALL FUNCTIONS IN SCHEMA public '
+                    'FROM ag_wwwuser CASCADE')
+
+        # ag schema
+        click.echo('Revoking all privileges in schema ag')
+        cur.execute('REVOKE ALL PRIVILEGES ON SCHEMA ag FROM ag_wwwuser '
+                    'CASCADE')
+
+        click.echo('Revoking all privileges on all tables in ag')
+        cur.execute('REVOKE ALL PRIVILEGES ON ALL TABLES IN SCHEMA ag '
+                    'FROM ag_wwwuser CASCADE')
+
+        click.echo('Revoking all privileges on all functions in ag')
+        cur.execute('REVOKE ALL PRIVILEGES ON ALL FUNCTIONS IN SCHEMA ag '
+                    'FROM ag_wwwuser CASCADE')
+    except:
+        # Some of these might not actually exist, just pass in that case
+        pass
+    finally:
         cur.close()
         conn.close()
 
+    conn = connect(user=AMGUT_CONFIG.user, password=AMGUT_CONFIG.password,
+                   host=AMGUT_CONFIG.host, port=AMGUT_CONFIG.port)
+    cur = conn.cursor()
+
+    cur.execute('SET SEARCH_PATH TO ag, public')
+
+    conn.set_isolation_level(ISOLATION_LEVEL_AUTOCOMMIT)
+
+    click.echo('Dropping user ag_wwwuser...')
+    cur.execute('DROP USER IF EXISTS ag_wwwuser')
+
+    click.echo('Dropping schema ag...')
+    cur.execute('DROP SCHEMA IF EXISTS ag CASCADE')
+
+    cur.close()
+    conn.close()
+
+    click.echo('Dropping database {}'.format(AMGUT_CONFIG.database))
 
-@env.command()
+    command = ['dropdb', '--if-exists']
```

Comment:
Ah, good point, that makes sense. Thanks!

On Fri, Jan 2, 2015 at 8:27 PM, Beau Gunderson <notifications@github.com>
wrote:

> In scripts/ag
> <https://github.com/biocore/american-gut-web/pull/354#discussion-diff-22429169>
> :
>
> >
> > -@env.command()
> > +    command = ['dropdb', '--if-exists']
>
> This was done early on as I was still learning how the automatic
> connection code worked; I thought a connection was being held open (and
> that was preventing the database from being dropped). After I refactored to
> use the lazy-loaded connections that issue went away, but I think this did
> fix another issue in my system:
>
> I setup my database so that the amgut user only had access to the amgut
> database. To delete the database via psycopg2 I don't think you can specify
> the database to connect to (since you'll then be prevented from dropping
> it). There are no other databases my user had access to from which to
> delete the database, but deleting with dropdb did work.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/american-gut-web/pull/354/files#r22429169>.
>